dashboard/lib/level_loader.rb

```diff
@@ -1,29 +1,98 @@
-class LevelLoader
-  def self.for_each_file(path, &block)
-    Dir.glob(Rails.root.join(path)).sort.map(&block)
-  end
+require 'set'
 
+class LevelLoader
+  # Top-level entry point, called by rake seed:custom_levels
   def self.load_custom_levels
-    level_index = Level.includes(:game).to_a.index_by(&:name)
-    for_each_file('config/scripts/**/*.level') do |path|
-      load_custom_level(path, level_index)
-    end
+    import_levels 'config/scripts/**/*.level'
   end
 
-  def self.level_file_path(name)
-    level_paths = Dir.glob(Rails.root.join("config/scripts/**/#{name}.level"))
-    raise("Multiple .level files for '#{name}' found: #{level_paths}") if level_paths.many?
-    level_paths.first || Rails.root.join("config/scripts/levels/#{name}.level")
+  #
+  # Loads a group of level files from disk and imports them into the database.
+  #
+  # - Level files not found in the database will be created.
+  # - Level files found in the database will be updated if they don't match the
+  #   file as loaded from disk.
+  #
+  # @param [String] level_file_glob - dashboard-relative, wildcard-friendly path
+  #   to one or more .level files.
+  #   Examples:
+  #     'config/scripts/levels/K-1 Bee 2.level'
+  #     'config/scripts/**/*.level'
+  #
+  def self.import_levels(level_file_glob)
+    level_file_paths = file_paths_from_glob(level_file_glob)
+
+    # Use a transaction because loading levels requires two separate imports.
+    Level.transaction do
+      level_md5s_by_name = Hash[Level.pluck(:name, :md5)]
+      existing_level_names = level_md5s_by_name.keys.to_set
+
+      # First, save stubs of any new levels - they'll need to have ids in
+      # order to create certain associations (in particular
+      # level_concept_difficulty) when we bulk-load the level properties.
+      new_level_names = level_file_paths.
+        map {|path| level_name_from_path path}.
+        reject {|name| existing_level_names.include? name}
+      Level.import! new_level_names.map {|name| {name: name}}
+
+      # Load level properties from disk and build a collection of levels that
+      # have changed.
+      changed_levels = level_file_paths.
+          map {|path| load_custom_level path, level_md5s_by_name}.
+          compact.
+          select(&:changed?)
+
+      # activerecord-import (with MySQL, anyway) doesn't save associated
+      # models, so we've got to do this manually.
+      changed_lcds = changed_levels.map(&:level_concept_difficulty).compact
+      lcd_update_columns = LevelConceptDifficulty.columns.map(&:name).map(&:to_sym).
+          reject {|column| %i{id level_id created_at}.include? column}
+      LevelConceptDifficulty.import! changed_lcds, on_duplicate_key_update: lcd_update_columns
```

Comment:
Are you sure that saving associated models doesn't work? My earlier commits seemed to have some logic related to eager-loading associated models which suggests that it might work with some slight tweaking ( https://github.com/code-dot-org/code-dot-org/compare/activerecord-import#diff-c6beca85a2968aae3fbc50912bb26afeR413): https://github.com/code-dot-org/code-dot-org/blob/de552a0809f64ea4bdfabd337af2693482acddb0/dashboard/app/models/script.rb#L413

Also note the added `inverse_of` lines in the branch, and also the added `accepts_nested_attributes_for :ideal_level_source`: https://github.com/code-dot-org/code-dot-org/blob/de552a0809f64ea4bdfabd337af2693482acddb0/dashboard/app/models/level.rb#L31

I don't remember if any of those actually worked or if they were just failed attempts, but it's at least worth testing if they might offer any improvement (whether performance or code-maintainability).

Reply:
Hmm, totally possible I thought this wasn't working before I realized `LevelConceptDifficulty` attributes not saving [was a totally separate pre-existing bug](https://github.com/code-dot-org/code-dot-org/pull/24034).  I'll tear out this custom logic and see if things still work.

Reply:
Tried this: Removed the separate import for `LevelConceptDifficulty`, ensured `inverse_of` was defined on the relevant associations, added `recursive: true` to the import command.  No joy - my unit tests suggest no `LevelConceptDifficulty` rows are created.

According to their wiki, [recursive imports are only supported by the PostgreSQL adapter](https://github.com/zdennis/activerecord-import/wiki#recursive-example-postgresql-only).  Browsing closed issues I'm seeing the same thing.